biz.aQute.tester.junit-platform/src/aQute/tester/bundle/engine/discovery/BundleSelectorResolver.java

```diff
@@ -0,0 +1,469 @@
+							.append("bundle", bundleDesc),
+						displayNameOf(bundle), new JUnitException("Fragment was not attached to a host bundle")));
+				});
+		}
+
+		Stream.of(allBundles)
+			.filter(isATestBundle)
+			.filter(BundleUtils::isFragment)
+			.filter(BundleUtils::isNotAttached)
+			.forEach(bundle -> {
+				final String bundleDesc = uniqueIdOf(bundle);
+				UniqueId bundleId = uniqueId.append("bundle", bundleDesc);
+				BundleDescriptor bd = new BundleDescriptor(bundle, bundleId,
+					new BundleException("Test fragment was not attached to a host bundle"));
+				b.addChild(bd);
+				markClassesResolved(bundle);
+			});
+
+		Stream.of(allBundles)
+			.filter(BundleUtils::isResolved)
+			.filter(BundleUtils::isNotFragment)
+			.forEach(bundle -> {
+				addEnginesToBundle(bundle);
+			});
+
+		Stream.of(allBundles)
+			.filter(isATestBundle)
+			.filter(BundleUtils::isFragment)
+			.filter(BundleUtils::isAttached)
+			.forEach(bundle -> {
+				info(() -> "Attached test fragment: " + bundle);
+				Bundle parent = BundleUtils.getHost(bundle)
+					.get();
+				BundleDescriptor bd = bundleMap.get(parent.getBundleId());
+				if (bd == null) {
+					UniqueId parentId = uniqueId.append("bundle", uniqueIdOf(parent));
+					bd = new BundleDescriptor(parent, parentId);
+					b.addChild(bd);
+					bundleMap.put(parent.getBundleId(), bd);
+				}
+				BundleDescriptor fd = new BundleDescriptor(bundle, bd.getUniqueId()
+					.append("fragment", uniqueIdOf(bundle)), null);
+				bd.addChild(fd);
+				addEnginesToBundleDescriptor(fd);
+			});
+
+		Stream.of(misconfiguredEnginesDescriptor, unresolvedBundlesDescriptor, unattachedFragmentsDescriptor)
+			.filter(StaticFailureDescriptor::hasChildren)
+			.forEach(b::addChild);
+
+		if (engines.isEmpty()) {
+			TestDescriptor t = new StaticFailureDescriptor(uniqueId.append("test", "noEngines"), "Initialization Error",
+				new JUnitException("Couldn't find any registered TestEngines"));
+			b.addChild(t);
+			return b;
+		}
+
+		if (testUnresolved && !unresolvedClasses.isEmpty()) {
+			StaticFailureDescriptor unresolvedClassesDescriptor = new StaticFailureDescriptor(
+				uniqueId.append("test", "unresolvedClasses"), "Unresolved classes");
+			for (String unresolvedClass : unresolvedClasses) {
+				unresolvedClassesDescriptor.addChild(new StaticFailureDescriptor(
+					unresolvedClassesDescriptor.getUniqueId()
+						.append("test", unresolvedClass),
+					unresolvedClass, new JUnitException("Couldn't find class " + unresolvedClass + " in any bundle")));
+			}
+			b.addChild(unresolvedClassesDescriptor);
+		}
+		info(() -> dump(b, ""));
+		return b;
+	}
+
+	void info(Supplier<String> msg, Throwable cause) {
+		if (verbose) {
+			System.err.println(msg.get());
+			cause.printStackTrace();
+		}
+	}
+
+	void info(Supplier<String> msg) {
+		if (verbose) {
+			System.err.println(msg.get());
+		}
+	}
+
+	private static Stream<String> testCases(Bundle bundle) {
+		final String testClasses = bundle.getHeaders()
```

Comment:
This can be more simply and efficiently written as 
```java
return aQute.lib.strings.Strings.splitAsStream(bundle.getHeaders()
	.get(aQute.bnd.osgi.Constants.TESTCASES));
```
You will need to add `aQute.libg;version=project` to `-buildpath`.

There are a lot of useful support functions in aQute.libg.

Reply:
Done, thanks for the tip.

I can see that there are many useful support functions in `aQute.libg`. It's just a bit hard to know to go looking for them if you're not already intimately aware of what's available and what it does. I suppose I should take the time to familiarize myself with them all. :smile: